iOS Documentation Revamp (#650)
ParsePlatform/Docs: _includes/ios/objects.md

## The PFObject

> Storing data on Parse is built around the `PFObject`. Each `PFObject` contains key-value pairs of JSON-compatible data. This data is schemeless, which means that you don't need to specify ahead of time what keys exist on each `PFObject`. You simply set whatever key-value pairs you want, and our backend will store it.

Reviewer: I think schemaless was correct here.

Author: Drrp. You're right.

Reviewer: This one still needs changing

There are two things to note here. You didn't have to configure or set up a new Class called `GameScore` before running this code. Your Parse app lazily creates this Class for you when it first encounters it.

> There are also a few fields you don't need to specify that are provided and set by the system as a convenience. `objectId` is a unique identifier for each saved object. `createdAt` and `updatedAt` represent the time that each object was created or last modified and saved to the Parse Server. Each of these fields is filled in by Parse, so they don't exist on a `PFObject` until the first save operation has been completed.

Reviewer: ```suggestion
There are also a few fields you don't need to specify that are provided and set by the system as a convenience. `objectId` is a unique identifier for each saved object. `createdAt` and `updatedAt` represent the time that each object was created or last modified and saved to the Parse Server. Each of these fields is filled in by Parse Server, so they don't exist on a `PFObject` until the first save operation has been completed.
```

```objective_c
PFQuery *query = [PFQuery queryWithClassName:@"GameScore"];
[query getObjectInBackgroundWithId:@"xWMyZ4YEGZ" block:^(PFObject *gameScore, NSError *error) {
    if (!error) {
        // Success!
    } else {
        // Failure!
    }
}];
```
```swift
let query = PFQuery(className:"GameScore")

> query.getObjectInBackground(withId: "xWMyZEGZ") { (object, error) in

Reviewer: The objective-c example still returns `gameScore` and the rest of the guide uses `gameScore` extensively so I think it is probably best to stick with that.

```suggestion
query.getObjectInBackground(withId: "xWMyZEGZ") { (gameScore, error) in
```

</div>

If you need to refresh an object you already have with the latest data that

> is in the cloud, you can use the `fetchInBackgroundWithBlock:` or `fetchInBackgroundWithTarget:selector:` methods to provide additional logic which will run after fetching the object.

Reviewer: ```suggestion
is in the database, you can use the `fetchInBackgroundWithBlock:` or `fetchInBackgroundWithTarget:selector:` methods to provide additional logic which will run after fetching the object.
```

> is in the database, you can use the `fetchInBackgroundWithBlock:` or `fetchInBackgroundWithTarget:selector:` methods to provide additional logic which will run after fetching the object.

Reviewer: As you've removed the `fetch` example I don't see the point in mentioning the benefit of the `withBlock` method...

```suggestion
is in the database, you can use the `fetchInBackgroundWithBlock:` or `fetchInBackgroundWithTarget:selector:` methods.
```

```
</div>

Note that it is not currently possible to atomically add and remove items from an array in the same save using. You will have to call `save` in between every different kind of array operation.

## Deleting Objects

> There are a few ways to delete a  `PFObject`.  For basic asynchronous deletion of a single object call the objects `deleteInBackground` function. If you prefer to recieve a callback you can use the `deleteInBackgroundWithBlock:` or `deleteInBackgroundWithTarget:selector:` methods. If you want to block the calling thread, you can use the `delete` method. Lastly,  `deleteEventually`  network consious solution that deletes when possible but does not guarantee a timeframe for the tasks completion.

Reviewer: ```suggestion

There are a few ways to delete a `PFObject`. For basic asynchronous deletion of a single object call the objects `deleteInBackground` function. If you prefer to recieve a callback you can use the `deleteInBackgroundWithBlock:` or `deleteInBackgroundWithTarget:selector:` methods. If you want to block the calling thread, you can use the `delete` method. Lastly, `deleteEventually` network conscious solution that deletes when possible but does not guarantee a timeframe for the tasks completion.
```

> For deleting multiple objects use the `PFObject` static function `deleteAllInBackground` to delete an array of objects asynchronous. The same can be done while blocking the calling thread using `deleteAll`. Lastly, to recieve a callback after deleting objects asyncronously use `deleteAllInBackground:block:` as demonstrated below.

Reviewer: ```suggestion
For deleting multiple objects use the `PFObject` static function `deleteAllInBackground` to delete an array of objects asynchronously. The same can be done while blocking the calling thread using `deleteAll`. Lastly, to recieve a callback after deleting objects asyncronously use `deleteAllInBackground:block:` as demonstrated below.
```

Author: God I feel so stupid for missing stupid stuff like spelling. I guess after about 5+ hours at it last night I started getting a bit sloppy. Theres just so much that needs to be done.

```
</div>

Note that it is not currently possible to atomically add and remove items from an array in the same save using. You will have to call `save` in between every different kind of array operation.

## Deleting Objects

> There are a few ways to delete a `PFObject`. For basic asynchronous deletion of a single object call the objects `deleteInBackground` function. If you prefer to recieve a callback you can use the `deleteInBackgroundWithBlock:` or `deleteInBackgroundWithTarget:selector:` methods. If you want to block the calling thread, you can use the `delete` method. Lastly, `deleteEventually` network conscious solution that deletes when possible but does not guarantee a timeframe for the tasks completion.

Reviewer: ```suggestion
There are a few ways to delete a `PFObject`. For basic asynchronous deletion of a single object call the objects `deleteInBackground` function. If you prefer to recieve a callback you can use the `deleteInBackgroundWithBlock:` or `deleteInBackgroundWithTarget:selector:` methods. If you want to block the calling thread, you can use the `delete` method. Lastly, `deleteEventually` is a network conscious option that deletes when possible but does not guarantee a timeframe for the tasks completion.
```

> For deleting multiple objects use the `PFObject` static function `deleteAllInBackground` to delete an array of objects asynchronously. The same can be done while blocking the calling thread using `deleteAll`. Lastly, to recieve a callback after deleting objects asyncronously use `deleteAllInBackground:block:` as demonstrated below.

Reviewer: Why have you completely removed reference of single object delete methods?

Author: I didnt. In fact its mentioned in the change you suggested above. Its just the way that github is the presenting the changes made.  What I did is take this:

`
>To delete an object from the cloud:

><div class="language-toggle" markdown="1">
>objective_c 
>[gameScore deleteInBackground];

>swift
>gameScore.deleteInBackground()

></div>

>If you want to run a callback when the delete is confirmed, you can use the deleteInBackgroundWithBlock: or deleteInBackgroundWithTarget:selector: methods. If you want to block the calling thread, you can use the delete method.`

and turn it into the the section you made a suggestion on above . Then I added additional information that was not documented before about PFObjects static functions for deleting arrays as you see above as well as an example for it.

Reviewer: Ah, my bad

PFUser *user = [PFUser currentUser];
PFRelation *relation = [user relationForKey:@"likes"];
[relation addObject:post];
[user saveInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
    if (succeeded) {
        // The post has been added to the user's likes relation.
    } else {
        // There was a problem, check error.description
    }
}];
```
```swift
guard let user = PFUser.current() else { return }
let relation = user.relation(forKey: "likes")
relation.add(post)
user.saveInBackground { (succeeded, error) in
    if (succeeded) {
        // The post has been added to the user's likes relation.
    } else {
        // There was a problem, check error.description
    }
}
```
</div>

> You can remove a post from the `PFRelation` similarly using the `removeObject:` function followed by saving the parent object.

Reviewer: Are you planning to favour the swift methods signatures as we discussed? or leaving this to another PR?

Author: Unfortunately, by the time I thought about it and got your input these changes had already been committed.

Reviewer: Ok, maybe we’ll leave this for a future pr?

Author: For sure. My plan is to kind of get through the rest of this and then return to update the docs with the swift functions as well as switch swift to the default example. Ill be doing this for the rest of the docs from here on out though.

Reviewer: Sounds good 👍

<div class="language-toggle" markdown="1">
```objective_c
[[relation query] findObjectsInBackgroundWithBlock:^(NSArray * _Nullable objects, NSError * _Nullable error) {
    if (error) {
        // There was an error
    } else {
        // objects has all the Posts the current user liked.
    }
}];
```
```swift
relation.query().findObjectsInBackground { (object, error) in
    if error == nil {
        // Success
    } else {
        // Failure!
    }
})
```
</div>

You can add constraints to a `PFRelation`'s query by adding constraints to the `PFQuery` returned by its `query` parameter as demonstrated below:

<div class="language-toggle" markdown="1">
```objective_c
PFQuery *query = [relation query];
[query whereKey:"category" equalTo:@"development"];
PFObject *object = [query getFirstObject]; // Query first object found
if (object) {
    // Do something with object
}
```
```swift
var query = relation.query()
query.whereKey("category", equalTo: "development") 

// Query first object found
if let object = try? query.getFirstObject() {
    // Do something with object
} ```
</div>

> You can learn more queries by visiting the [PFQuery](#queries) section.  A `PFRelation` behaves similarly to an array of  `PFObject` yet has a built in query that's capable of everything a standard `PFQuery` is other than `includeKey:`.

Reviewer: ```suggestion
You can learn more about queries by visiting the [PFQuery](#queries) section. A `PFRelation` behaves similarly to an array of  `PFObject` yet has a built in query that's capable of everything a standard `PFQuery` is other than `includeKey:`.
```

  PFImageView *view = [[PFImageView alloc] initWithImage:kPlaceholderImage];
  view.file = self.iconFile;
  [view loadInBackground];
  
  return view;
}
```

```swift
@NSManaged var iconFile: PFFileObject!

func iconView() -> UIImageView {
  let view = PFImageView(imageView: PlaceholderImage)
  view.file = iconFile
  view.loadInBackground()
  
  return view
}
```
</div>

### Initializing Subclasses

> You should initialize new instances of subclassses with standard initialization methods. To create a new instance to an existing Parse object, use the inherited `PFObject` class function `objectWithoutDataWithObjectId:`, or create a new object and set the objectId property manually.

Reviewer: Not sure this sentence makes sense...

```suggestion
You should initialize new instances of subclassses with standard initialization methods. To create a new instance of an existing Parse object, use the inherited `PFObject` class function `objectWithoutDataWithObjectId:`, or create a new object and set the objectId property manually.
```
